### PR TITLE
[Vec256][neon] Add neon backend for vec256

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40252 Add add_relu fusion pass to optimize_for_mobile.
* #39343 JIT pass for add relu fusion.
* #39342 Add fused add_relu op.
* **#39341 [Vec256][neon] Add neon backend for vec256**
* #40059 Add benchmark for add op.

Summary:
This PR introduces neon backend for vec256 class for float datatype.
For now only aarch64 is enabled due to few issues with enabling in
aarch32 bit.

Test Plan:
vec256_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822399](https://our.internmc.facebook.com/intern/diff/D21822399)

### CODE REVIEW

**Comment by kimishpatel** [comment]
Tensor Add benchmark: c = a + b (a, b are tensors).

Before this PR:
---------------
--------------------------------------------------------------
Benchmark                       Time           CPU Iterations
--------------------------------------------------------------
tensor_add/N:8/C:8           1896 ns       1881 ns     374010
tensor_add/N:8/C:16          1911 ns       1907 ns     358576
tensor_add/N:8/C:32          2068 ns       2064 ns     340726
tensor_add/N:8/C:64          2490 ns       2465 ns     296849
tensor_add/N:8/C:128         3137 ns       3127 ns     224866
tensor_add/N:8/C:256         4514 ns       4503 ns     154018
tensor_add/N:8/C:512         7457 ns       7434 ns      95589
tensor_add/N:16/C:8          1921 ns       1912 ns     362732
tensor_add/N:16/C:16         2123 ns       2119 ns     331399
tensor_add/N:16/C:32         2511 ns       2464 ns     285673
tensor_add/N:16/C:64         3014 ns       3009 ns     226860
tensor_add/N:16/C:128        4507 ns       4486 ns     157992
tensor_add/N:16/C:256        7789 ns       7761 ns      91265
tensor_add/N:16/C:512       14375 ns      14326 ns      49529
tensor_add/N:32/C:8          2116 ns       2111 ns     330518
tensor_add/N:32/C:16         2384 ns       2379 ns     295629
tensor_add/N:32/C:32         3096 ns       3090 ns     232728
tensor_add/N:32/C:64         4555 ns       4545 ns     152578
tensor_add/N:32/C:128        7696 ns       7673 ns      90817
tensor_add/N:32/C:256       13716 ns      13682 ns      50496
tensor_add/N:32/C:512       50113 ns      49814 ns      14586
tensor_add/N:64/C:8          2451 ns       2447 ns     280541
tensor_add/N:64/C:16         3181 ns       3133 ns     223684
tensor_add/N:64/C:32         4375 ns       4368 ns     157079
tensor_add/N:64/C:64         7416 ns       7400 ns      95118
tensor_add/N:64/C:128       14208 ns      14168 ns      50237
tensor_add/N:64/C:256       53626 ns      52982 ns      14424
tensor_add/N:64/C:512      101495 ns     101195 ns       7081
tensor_add/N:128/C:8         3017 ns       3013 ns     228757
tensor_add/N:128/C:16        4567 ns       4556 ns     151779
tensor_add/N:128/C:32        7721 ns       7700 ns      91759
tensor_add/N:128/C:64       14274 ns      14135 ns      49453
tensor_add/N:128/C:128      55285 ns      55065 ns      10000
tensor_add/N:128/C:256      96954 ns      96257 ns       6053
tensor_add/N:128/C:512     132551 ns     131370 ns       5482
tensor_add/N:256/C:8         4482 ns       4472 ns     156483
tensor_add/N:256/C:16        7825 ns       7796 ns      91233
tensor_add/N:256/C:32       13863 ns      13832 ns      50097
tensor_add/N:256/C:64       51234 ns      51052 ns      11562
tensor_add/N:256/C:128      94868 ns      94346 ns       7586
tensor_add/N:256/C:256     172585 ns     171331 ns       4696
tensor_add/N:256/C:512     253899 ns     251875 ns       2673
tensor_add/N:512/C:8         7906 ns       7875 ns      86695
tensor_add/N:512/C:16       13617 ns      13593 ns      48361
tensor_add/N:512/C:32       50016 ns      49825 ns      14081
tensor_add/N:512/C:64      100958 ns     100413 ns       6734
tensor_add/N:512/C:128     136364 ns     135252 ns       4703
tensor_add/N:512/C:256     244822 ns     243009 ns       2761
tensor_add/N:512/C:512     636578 ns     629219 ns       1314
----------------------------------------------------------------------


After this PR:
--------------------------------------------------------------
Benchmark                       Time           CPU Iterations
--------------------------------------------------------------
tensor_add/N:8/C:8           1823 ns       1820 ns     376542
tensor_add/N:8/C:16          1894 ns       1870 ns     377604
tensor_add/N:8/C:32          2022 ns       2006 ns     358275
tensor_add/N:8/C:64          2131 ns       2127 ns     328004
tensor_add/N:8/C:128         2345 ns       2340 ns     293971
tensor_add/N:8/C:256         2857 ns       2841 ns     248697
tensor_add/N:8/C:512         4134 ns       4100 ns     175434
tensor_add/N:16/C:8          1925 ns       1920 ns     365383
tensor_add/N:16/C:16         1992 ns       1986 ns     354167
tensor_add/N:16/C:32         2097 ns       2091 ns     328462
tensor_add/N:16/C:64         2320 ns       2316 ns     305146
tensor_add/N:16/C:128        4495 ns       4474 ns     241114
tensor_add/N:16/C:256        4018 ns       4011 ns     173206
tensor_add/N:16/C:512        6405 ns       6391 ns     104671
tensor_add/N:32/C:8          1914 ns       1911 ns     363808
tensor_add/N:32/C:16         2146 ns       2137 ns     342782
tensor_add/N:32/C:32         2393 ns       2388 ns     294595
tensor_add/N:32/C:64         2951 ns       2943 ns     233958
tensor_add/N:32/C:128        3935 ns       3921 ns     173340
tensor_add/N:32/C:256        6386 ns       6363 ns     108047
tensor_add/N:32/C:512       34577 ns      34389 ns      20863
tensor_add/N:64/C:8          2151 ns       2145 ns     328347
tensor_add/N:64/C:16         2331 ns       2326 ns     293876
tensor_add/N:64/C:32         2873 ns       2862 ns     245913
tensor_add/N:64/C:64         4112 ns       4094 ns     171724
tensor_add/N:64/C:128        6596 ns       6579 ns     105594
tensor_add/N:64/C:256       34590 ns      34462 ns      20719
tensor_add/N:64/C:512       71910 ns      71617 ns       9807
tensor_add/N:128/C:8         2355 ns       2348 ns     302543
tensor_add/N:128/C:16        2931 ns       2925 ns     237986
tensor_add/N:128/C:32        4143 ns       4131 ns     171819
tensor_add/N:128/C:64        6433 ns       6415 ns     110009
tensor_add/N:128/C:128      33633 ns      33541 ns      17884
tensor_add/N:128/C:256      73210 ns      72819 ns      10946
tensor_add/N:128/C:512     139340 ns     138525 ns       4630
tensor_add/N:256/C:8         2941 ns       2933 ns     237760
tensor_add/N:256/C:16        4027 ns       4018 ns     169902
tensor_add/N:256/C:32        6405 ns       6391 ns     105715
tensor_add/N:256/C:64       41040 ns      40635 ns      22174
tensor_add/N:256/C:128      74922 ns      74683 ns       8029
tensor_add/N:256/C:256     126337 ns     122638 ns       6762
tensor_add/N:256/C:512     208467 ns     207109 ns       3318
tensor_add/N:512/C:8         3988 ns       3977 ns     175640
tensor_add/N:512/C:16        6615 ns       6546 ns     110629
tensor_add/N:512/C:32       36946 ns      36852 ns      17129
tensor_add/N:512/C:64       68387 ns      68110 ns      10976
tensor_add/N:512/C:128     118401 ns     117374 ns       6842
tensor_add/N:512/C:256     216552 ns     214377 ns       2426
tensor_add/N:512/C:512     484803 ns     478878 ns       1462

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+          a.values.val[1], b.values.val[1], vec.values.val[1]);
+    return vec;
+  }
+  static Vec256<float> blendv(const Vec256<float>& a, const Vec256<float>& b,
+                              const Vec256<float>& mask) {
+    // TODO
+    // NB: This requires that each value, i.e., each uint value,
+    // of the mask either all be zeros or all be 1s.
+    // We perhaps need some kind of an assert?
+    // But that will affect performance.
```
Comment (AshkanAliabadi) [inline aten/src/ATen/cpu/vec256/vec256_float_neon.h:117]: Asserts should be compiled out in release builds.  Having a check in debug builds is worthwhile I would say.

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 2]
```diff
@@ -0,0 +1,653 @@
...
+        vreinterpretq_u32_f32(vec.values.val[1]),
+        b.values.val[1],
+        a.values.val[1]);
+    return vec;
+  }
+  template<typename step_t>
+  static Vec256<float> arange(float base = 0.f, step_t step = static_cast<step_t>(1)) {
+    return Vec256<float>(
+      base,            base +     step, base + 2 * step, base + 3 * step,
+      base + 4 * step, base + 5 * step, base + 6 * step, base + 7 * step);
```
Comment (AshkanAliabadi) [inline aten/src/ATen/cpu/vec256/vec256_float_neon.h:133]: This could be vectorized itself.  Currently all these operations are done on scalars.  Imagine _base_, _i_ = [0, 1, 2, 3, 4, 5, 6, 7], and _step_ are vectors, then you could rewrite as _base_ + _i_ * _step_.

**Review by AshkanAliabadi — COMMENTED** [review]
Will review the rest later.

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+public:
+  using value_type = float;
+  static constexpr int size() {
+    return 8;
+  }
+  Vec256() {}
+  Vec256(float32x4x2_t v) : values(v) {}
+  Vec256(float val) {
+    values.val[0] = vdupq_n_f32(val);
+    values.val[1] = vdupq_n_f32(val);
```
[the remark this answers is not in the document]
Comment (kimishpatel) [reply aten/src/ATen/cpu/vec256/vec256_float_neon.h:66]: Sounds good.

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+  Vec256(float32x4_t val0, float32x4_t val1) {
+    values.val[0] = val0;
+    values.val[1] = val1;
+  }
+  operator float32x4x2_t() const {
+    return values;
+  }
+  template <int64_t mask>
+  static Vec256<float> blend(const Vec256<float>& a, const Vec256<float>& b) {
+    Vec256<float> vec;
```
[the remark this answers is not in the document]
Comment (kimishpatel) [reply aten/src/ATen/cpu/vec256/vec256_float_neon.h:82]: True. Let me think about it. It is little unreadable and verbose. Although, https://en.cppreference.com/w/cpp/language/copy_elision, says that such cases should be copy elided although they are not mandatory unlike the one you proposed which is mandatory by cpp standard.

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/cpu/vec256/vec256_float_neon.h:117]: Do you mean TORCH_INTERNAL_ASSERT?

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/cpu/vec256/vec256_float_neon.h:133]: Good point.

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (AshkanAliabadi) [reply aten/src/ATen/cpu/vec256/vec256_float_neon.h:82]: Yes my suggestions may be antiquated.  There's NRVO, and also a mandatory copy elision introduced in C++17 so I am not sure whether this practice is still of value or not.

**Comment by AshkanAliabadi** [comment]
Out of curiosity, how do you read the benchmark results?  For instance for

tensor_add/N:32/C:256 13716 ns 13682 ns 50496

I can tell it's for batch size of 32 with 256 channels but why are three timings instead of one?  Max (?), min (?), and what's the third one?

**Comment by kimishpatel** [comment]
> Out of curiosity, how do you read the benchmark results? For instance for
> 
> tensor_add/N:32/C:256 13716 ns 13682 ns 50496
> 
> I can tell it's for batch size of 32 with 256 channels but why are three timings instead of one? Max (?), min (?), and what's the third one?

First is wallclock time I think. Second is cpu time and third is number iterations it was run for. Looking at the doc it seems that the number of iterations is governed by some min time for cpu and wallclock that benchmark needs to hit.

File: CMakeLists.txt [hunk 5]
```diff
@@ -133,6 +133,7 @@ cmake_dependent_option(
     "NOT BUILD_SHARED_LIBS" OFF)
 option(BUILD_TEST "Build C++ test binaries (need gtest and gbenchmark)" OFF)
 option(BUILD_MOBILE_BENCHMARKS "Build C++ test binaries for mobile (ARM) targets(need gtest and gbenchmark)" OFF)
+option(BUILD_MOBILE_TEST "Build C++ test binaries for mobile (ARM) targets(need gtest and gbenchmark)" OFF)
```
Comment (dreiss) [inline CMakeLists.txt:136]: Space before `(`

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 6]
```diff
@@ -120,8 +120,7 @@ struct Vec256 {
   }
   template<typename... Args,
            typename = std::enable_if_t<(sizeof...(Args) == size())>>
-  Vec256(Args... vals) {
-    values = { vals... };
+  Vec256(Args... vals) : values{vals...}{
```
Comment (dreiss) [inline aten/src/ATen/cpu/vec256/vec256_base.h:123]: Revert unrelated change?

File: aten/src/ATen/cpu/vec256/vec256_qint.h [hunk 7]
```diff
@@ -1093,10 +1093,19 @@ struct Vec256QuantizedConverter {
...
             scale[j], zero_point[j], T(vals[8 * i + j]));
       }
+      rv[i] = Vec256<float>(tmp_vals[0],
+          tmp_vals[1],
+          tmp_vals[2],
+          tmp_vals[3],
+          tmp_vals[4],
+          tmp_vals[5],
+          tmp_vals[6],
+          tmp_vals[7]);
```
Comment (dreiss) [inline aten/src/ATen/cpu/vec256/vec256_qint.h:1108]: Why did this change?

File: caffe2/CMakeLists.txt [hunk 8]
```diff
@@ -1132,6 +1132,18 @@ if(BUILD_MOBILE_BENCHMARK)
   endforeach()
 endif()
 
+if(BUILD_MOBILE_TEST)
+  foreach(test_src ${ATen_MOBILE_TEST_SRCS})
```
Comment (dreiss) [inline caffe2/CMakeLists.txt:1136]: Does this need to be a separate thing?  How come this can't be part of the normal `BUILD_TEST`?

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+// However for now opting for STL, since we are not building
+// with Sleef for mobile yet.
+
+namespace at {
+namespace vec256 {
+// See Note [Acceptable use of anonymous namespace in header]
+namespace {
+
+// For 32 bit arm rely on neon for 64 bit advanced vector extension
+// Currently only suuporting aarch64 for two reasons.
```
Comment (dreiss) [inline aten/src/ATen/cpu/vec256/vec256_float_neon.h:19]: "suuporting"

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+// 1. For some reason arm_neon.h does not export all the intrinsics.
+//    Thought about using inline assembly for not supported ones.
+// 2. However android nkd r21 has problems with compiling aarch32.
```
Comment (dreiss) [inline aten/src/ATen/cpu/vec256/vec256_float_neon.h:22]: "NDK"

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 11]
```diff
@@ -0,0 +1,653 @@
...
+// such as sin, cos, tan etc..
+// However for now opting for STL, since we are not building
+// with Sleef for mobile yet.
+
+namespace at {
+namespace vec256 {
+// See Note [Acceptable use of anonymous namespace in header]
+namespace {
+
+// For 32 bit arm rely on neon for 64 bit advanced vector extension
```
Comment (dreiss) [inline aten/src/ATen/cpu/vec256/vec256_float_neon.h:18]: I don't understand this comment.  We're not supporting 32-bit at all right now, right?

**Review by dreiss — COMMENTED** [review]
I didn't review the full implementation of each op, but I'm willing to trust the test.  Ashkan, do you want to review, or should I just approve?

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 6 above)
Comment (kimishpatel) [reply aten/src/ATen/cpu/vec256/vec256_base.h:123]: Had to make this change because constructing values as before was not working. Mainly values being an array can only be initialized by `values{...}` and not by assignment, apparently. For some reason it worked before since there were no calls to this constructor in the codepath, but due to my change here, https://github.com/pytorch/pytorch/pull/39341/files#diff-a1ca038eb7df0b763436f68ddd635e3fR1101, now there is.

File: aten/src/ATen/cpu/vec256/vec256_qint.h (hunk 7 above)
Comment (kimishpatel) [reply aten/src/ATen/cpu/vec256/vec256_qint.h:1108]: So what happens is that the use of `Vec256QuantizedConverter` is only for non-AVX path, which is the case for us. This change is for dequantize util. Now Vec256<float> class that is not specialized for AVX provides `[]` operator here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/vec256_base.h#L131, so that you can effectively update the values held by Vec256<float> class directly. So the values held by vec256<float> type are accessed as a pointer and can be dereferenced. However Vec256<float> implementation for ARM, similar to AVX cannot provide a way to directly modify particular element of the vector datatype it holds by providing a pointer that can be dereferenced. That is why this change. And this leads to the change above.

File: caffe2/CMakeLists.txt (hunk 8 above)
Comment (kimishpatel) [reply caffe2/CMakeLists.txt:1136]: Mainly so that we can build only tests we specifically care for mobile instead of building all the tests during mobile build. But I am ok merging the two if you feel thats better.

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h (hunk 11 above)
Comment (kimishpatel) [reply aten/src/ATen/cpu/vec256/vec256_float_neon.h:18]: I have lost the context myself :). I will update the comment. Mainly I think I want to point out that for aarch32 we will do inline asm to implement all the corresponding functions.

**Comment by AshkanAliabadi** [comment]
> I didn't review the full implementation of each op, but I'm willing to trust the test. Ashkan, do you want to review, or should I just approve?

I'll take another pass by the end of the day.

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 6 above)
Comment (dreiss) [reply aten/src/ATen/cpu/vec256/vec256_base.h:123]: Oh, right.  I thought this was just an auto-format.  Didn't look closely enough. :)

File: caffe2/CMakeLists.txt (hunk 8 above)
Comment (dreiss) [reply caffe2/CMakeLists.txt:1136]: I would merge them.  If they're taking too long to build, maybe we could look at adding support for a `-D` flag that would specify a keeplist that would only build a subset of tests.

**Review by dreiss — COMMENTED** [review]
LGTM.  I'll leave to Ashkan to give the final approve.

File: caffe2/CMakeLists.txt (hunk 8 above)
Comment (kimishpatel) [reply caffe2/CMakeLists.txt:1136]: Actually I just realized, BUILD_TEST is hidden behind `NOT INTERN_BUILD_MOBILE`, https://github.com/pytorch/pytorch/blob/master/caffe2/CMakeLists.txt#L678. That is why I added this separate one. With all the different diffs I keep forgetting that I already tried looking into it before.

**Review by dreiss — APPROVED** [review]
